### ISSUE TITLE
F# SDK needs to import design-time targets and authored into VS

### ISSUE THREAD

**Issue by srivatsn** [dotnet/project-system#1857]
The Microsoft.FSharp.DesignTime.targets needs to be imported just before importing Microsoft.Common.targets in these locations (that I know of):

- FSharp.NET.SDK in the CLI repo - Sdk.OnRestore.targets
- FSharp.NET.SDK NuGet package - FSharp.NET.Current.Sdk.targets
- Microsoft.FSharp.targets under C:\Program Files (x86)\Microsoft SDKs\F#\4.1\Framework\v4.0, so that non-SDK projects know to import the design targets as well.

The code to import looks like this:
```xml
    <!-- Import design time targets for Roslyn Project System. These are only available if Visual Studio is installed. -->
    <!-- Import design time targets before the common targets, which import targets from Nuget. -->
    <PropertyGroup>
       <FSharpDesignTimeTargetsPath Condition="'$(FSharpDesignTimeTargetsPath)'==''">$(MSBuildExtensionsPath)\Microsoft\VisualStudio\Managed\Microsoft.FSharp.DesignTime.targets</FSharpDesignTimeTargetsPath>
    </PropertyGroup>
    <Import Project="$(FSharpDesignTimeTargetsPath)" Condition="'$(FSharpDesignTimeTargetsPath)' != '' and Exists('$(FSharpDesignTimeTargetsPath)')" />
```

The FSharp.NET.SDK from the CLI repo needs to be authored into VS as well. Ideally as part of this, the whole thing gets rationalized with the Microsoft.NET.Sdk.

**Comment by srivatsn** [comment]
@enricodasa @KevinRansom as well.

**Comment by enricosada** [comment]
> Ideally as part of this, the whole thing gets rationalized with the Microsoft.NET.Sdk.

with sdk 2.0, it should be possibile to reference sdk packages from nupkg, so no need, more easy to change it like that ihmo

> The FSharp.NET.SDK from the CLI repo needs to be authored into VS as well

vs should import the same Sdk bit the .net core sdk does https://github.com/dotnet/cli/blob/85ca206d84633d658d7363894c4ea9d59e515c1a/build/BundledSdks.props#L8

**Comment by srivatsn** [comment]
> with sdk 2.0, it should be possibile to reference sdk packages from nupkg
This is not the current plan for 2.0. The current plan is that VS will use the CLI's copy of the SDK and the acquisition story is to get a newer CLI. We still want to acquire SDKs trhough 

The reason I think we should have a real SDK instead of a packageref is that otherwise the VS experience tends to be poor. For C#, we started with the SDK as a packageref and when a project is created and before it's restored, it's in a weird state with wrong defaults for properties. For eg the web projects for C# would look like a console app with no wwwroot etc and then suddenly after some time, solution explorer would refresh and then the project will be usable. It was a very jarring experience and led to many many bugs. So we moved to a model where the project doesn't load if the SDK  is not present. I don't know how important the FSharp targets are but I suspect the project will be unusable with just the common.targets imported before restore.

**Comment by enricosada** [comment]
> This is not the current plan for 2.0. The current plan is that VS will use the CLI's copy of the SDK and the acquisition story is to get a newer CLI. We still want to acquire SDKs as packages but that's likely post 2.0.

sigh, that was a really nice feature ihmo. let's wait 3.0 for that, np 😄 
Yes, `FSharp.NET.Sdk` bundled in cli/VS should be updated with some new development. 
But atm is partial by design (is splitted) because like that i can update it just releasing a new FSharp.NET.Sdk package (without waitining a new cli release), because there are some wip/bugfix todo.
That is really good, i fixed some bugs, people use it, and i can for example fix it for your usage, without waiting for a new cli release.

FSharp.target contribute just `CoreCompile` target pratically, and is integrated in the right place atm by design, because is imported by Common.Target using a `LanguageTargets` property (the same used by c#) defined in restored `Fsharp.NET.Sdk`, is not randomly imported (who can create issues).

I need to add the `DefaultProjectTypeGuid `, `EnableDefaultCompileItems` properites to the embedded `Fsharp.NET.SDK`, that's for sure, so are ready before the restore, and project can be loaded without any issue.

**Comment by enricosada** [comment]
@srivatsn but i need to add 

```xml
       <FSharpDesignTimeTargetsPath Condition="'$(FSharpDesignTimeTargetsPath)'==''">$(MSBuildExtensionsPath)\Microsoft\VisualStudio\Managed\Microsoft.FSharp.DesignTime.targets</FSharpDesignTimeTargetsPath>
```

too? is not always passed from VS that prop? having hardcoded stuff (ok relative) in props is always annoying

is not 

```xml
<Import Project="$(FSharpDesignTimeTargetsPath)" Condition="'$(FSharpDesignTimeTargetsPath)' != '' and Exists('$(FSharpDesignTimeTargetsPath)')" />
```

enough?

**Comment by enricosada** [comment]
I'll prepare the `FSharp.NET.SDK` to embeed with DefaultProjectTypeGuid, EnableDefaultCompileItems and updated import in the targets like Sdk.OnRestore.targets
it's going to be useful also for https://github.com/dotnet/netcorecli-fsc/issues/56 because dotnet/cli asked a clean nupkg to bundle (instead of remove files)

**Comment by srivatsn** [comment]
No, FSharpDesignTimeTargetsPath is not set globally. So you'll need that line. The path is relative to msbuildextensionspath though - that's always set.

**Comment by srivatsn** [comment]
@enricosada @brettfo  I tried FSharp.NET.SDK 1.0.3 and this doesn't seem to be fixed yet? This is needed to open F# projects in VS.

**Comment by enricosada** [comment]
@srivatsn the `Sdk.OnRestore.targets` is not done yet. that is tracked in https://github.com/dotnet/netcorecli-fsc/issues/106
i am doing that now

**Comment by srivatsn** [comment]
Thanks @enricosada - just to make sure this will also be fixed in the FSharp.NET.Current.Sdk.targets in the Nuget package as well right? 1.0.3 doesn't seem to have that fixed.

**Comment by enricosada** [comment]
Thanks @srivatsn for checking, `1.0.3` fixed the compiler execution from msbuild.
I am doing this for roslyn in [v1.0.4](https://github.com/dotnet/netcorecli-fsc/milestone/14)
But thanks for `FSharp.NET.Current.Sdk.targets` remainder!
I'll publish asap, probably tonight.

building this repo atm, so i can check if works ok.

**Comment by enricosada** [comment]
@srivatsn @brettfo done, `1.0.4-*` prerelease packages are in a myget feed, can you or check? I tried locally and seems ok.
if ok for you, please comment in the PR, i'll publish these on nuget.org asap.
More info in https://github.com/dotnet/netcorecli-fsc/pull/111#issuecomment-298136538

/cc @KevinRansom @cartermp @dsyme

@brettfo i tried with this repo at commit b3c322789de4ab952d1c35485c53c076acd63800 and projects open, but intellisense doesnt start (i think is ok, doesn resolve fsc args yet). 

there are also projects ready in the PR Comment, but just use default `dotnet new -lang f#` templates with `<PackageReference Include="FSharp.NET.Sdk" Version="1.0.4-*" PrivateAssets="All" />` is enough

**Comment by srivatsn** [comment]
Thanks @enricosada - I tried 1.0.4 and it works fine for me.

**Comment by enricosada** [comment]
Awesome @srivatsn .
i sent a PR https://github.com/dotnet/cli/pull/6468 t to `dotnet/cli` (merged already) to embeed that `1.0.4-bundled-0100` in `Sdks`.
i think @cartermp said is enogh to be be authorerd inside VS bundled Sdks.

I hope can be embeeded AS IS also in https://github.com/mono/msbuild (/cc @radical @nosami ) to just have one source code, and others just embedd it :D

**Comment by srivatsn** [comment]
Yes VS is going to pick up the SDK from the CLI using a SDK resolver feature implemented in msbuild (https://github.com/Microsoft/msbuild/pull/2002/commits/b0a6d07124bb02e07808845ac36410778b34f0eb). The CLI drops a resolver into the desktop msbuild's directory. Ideally the same would happen for mono msbuild. /cc @livarcocc as well.

**Comment by enricosada** [comment]
Maybe also @radical and @nosami are interested to https://github.com/Microsoft/msbuild/pull/2002/commits/b0a6d07124bb02e07808845ac36410778b34f0eb for mono.

**Comment by radical** [comment]
I will merge from upstream again in a few days and will have that commit. This won't be included for 15.2 though!
